jsvg-javafx/src/main/java/com/github/weisj/jsvg/renderer/jfx/impl/FXOutput.java

```diff
@@ -0,0 +1,510 @@
+    }
+
+    @Override
+    public void translate(double dx, double dy) {
+        ctx.translate(dx, dy);
+    }
+
+    @Override
+    public float currentOpacity() {
+        return currentOpacity;
+    }
+
+    @Override
+    public void applyOpacity(float opacity) {
+        if (GeometryUtil.approximatelyEqual(opacity, 1)) return;
+        setOpacity(opacity * currentOpacity);
+    }
+
+    private void setOpacity(float opacity) {
+        currentOpacity = opacity;
+
+        // Re-apply paint with correct opacity
+        FXAWTBridge.applyPaint(ctx, currentPaint, currentOpacity);
+    }
+
+    @Override
+    public @NotNull SafeState safeState() {
+        return new FXOutputState(SaveClipStack.YES);
+    }
+
+    @Override
+    public boolean supportsFilters() {
+        return true;
+    }
+
+    @Override
+    public boolean supportsColors() {
+        return true;
+    }
+
+    @Override
+    public boolean isSoftClippingEnabled() {
+        // JavaFX performs soft clips by default
+        return false;
+    }
+
+    @Override
+    public boolean hasMaskedPaint() {
+        return currentPaint instanceof MaskedPaint;
+    }
+
+    private enum SaveClipStack {
+        YES,
+        NO
+    }
+
+    private class FXOutputState implements SafeState {
+
+        private final AffineTransform originalTransform;
+        private final Paint originalPaint;
+        private final Stroke originalStroke;
+        private final float originalOpacity;
+        private final List<Shape> originalClipStack;
+
+        public FXOutputState(SaveClipStack saveClip) {
+            this.originalTransform = transform();
+            this.originalPaint = currentPaint;
+            this.originalStroke = currentStroke;
+            this.originalOpacity = currentOpacity;
+            this.originalClipStack = saveClip == SaveClipStack.YES ? clipStack.snapshot() : null;
+        }
+
+        public @NotNull GraphicsContext context() {
+            return ctx;
+        }
+
+        @Override
+        public void restore() {
+            if (originalClipStack != null) {
+                clipStack.restoreClipStack(originalClipStack);
+            }
+            setOpacity(originalOpacity);
+            setTransform(originalTransform);
+            setPaint(originalPaint);
+            setStroke(originalStroke);
+        }
+    }
+
+    private static class ClipShape {
+
+        private final Shape shape;
+        private Rectangle2D bounds;
+        private final int savePoint; // Save point before the clip has been applied
+
+        private ClipShape(Shape shape, int savePoint) {
+            this.savePoint = savePoint;
+            this.shape = shape;
+        }
+
+        public Rectangle2D getBounds() {
+            if (bounds == null) {
+                // We need to apply the save points inverse transform to this
+                bounds = shape.getBounds2D();
+            }
+            return bounds;
+        }
+    }
+
+    private class ClipStack {
```

Comment:
I think this class does not actually need to store all `ClipShape`s as it only needs the sum of the `savePoints`s. Just accumulating these should actually be enough and reduce memory pressure from repeatedly pushing/popping from the underlying stack.

Reply:
I see the problem now. I think in the mesh gradient case it should also suffice to use `applyClip`. We could therefore remove all uses of `setClip` and deprecate it for removal in the future. The other assumptions should always hold currently and we can document these assumptions in the `Output` class.

The possibility to avoid having to keep the shapes around would be quite nice.